### Crimes in Chicago from 2001 to Present
### HYPOTHESIS
Learning what areas are affected by crime and seeing if crime has either gone up or down. With information like this we can def use it to our advantage in the sense of where help is needed. 

Patricia Jaquez
June 20, 2023

IMPORTS 

In [1]:
import pandas as pd
import pathlib as Path 
import plotly.express as px
import numpy as np
import matplotlib.pyplot as plt
import csv
from itertools import groupby


In [2]:
df = pd.read_csv(r'C:\Users\patty\OneDrive\Desktop\capstone_I\Crimes_-_2001_to_Present.csv')

In [3]:
df 

ID Case Number                    Date                  Block   
0        10224738    HY411648  09/05/2015 01:30:00 PM        043XX S WOOD ST  \
1        10224739    HY411615  09/04/2015 11:30:00 AM    008XX N CENTRAL AVE   
2        11646166    JC213529  09/01/2018 12:01:00 AM  082XX S INGLESIDE AVE   
3        10224740    HY411595  09/05/2015 12:45:00 PM      035XX W BARRY AVE   
4        10224741    HY411610  09/05/2015 01:00:00 PM    0000X N LARAMIE AVE   
...           ...         ...                     ...                    ...   
7784659  12847575    JF420478  09/01/2022 05:00:00 AM        005XX W SURF ST   
7784660  12847801    JF420319  07/08/2022 12:00:00 AM    114XX S PRAIRIE AVE   
7784661  12847324    JF420102  09/27/2022 11:00:00 AM        023XX E 70TH ST   
7784662  12847570    JF420427  09/03/2022 10:25:00 AM     052XX W CARMEN AVE   
7784663  12840464    JF411839  09/26/2022 07:20:00 PM      0000X N MASON AVE   

         IUCR        Primary Type                    Description   
0        0486             BATTERY        DOMESTIC BATTERY SIMPLE  \
1        0870               THEFT                 POCKET-PICKING   
2        0810               THEFT                      OVER $500   
3        2023           NARCOTICS          POSS: HEROIN(BRN/TAN)   
4        0560             ASSAULT                         SIMPLE   
...       ...                 ...                            ...   
7784659  2825       OTHER OFFENSE        HARASSMENT BY TELEPHONE   
7784660  1130  DECEPTIVE PRACTICE       FRAUD OR CONFIDENCE GAME   
7784661  0810               THEFT                      OVER $500   
7784662  2021           NARCOTICS         POSSESS - BARBITURATES   
7784663  143A   WEAPONS VIOLATION  UNLAWFUL POSSESSION - HANDGUN   

                    Location Description  Arrest  Domestic  ...  Ward   
0                              RESIDENCE   False      True  ...  12.0  \
1                                CTA BUS   False     False  ...  29.0   
2                              RESIDENCE   False      True  ...   8.0   
3                               SIDEWALK    True     False  ...  35.0   
4                              APARTMENT   False      True  ...  28.0   
...                                  ...     ...       ...  ...   ...   
7784659                        RESIDENCE   False      True  ...  44.0   
7784660                           STREET   False     False  ...   9.0   
7784661                        RESIDENCE   False     False  ...   5.0   
7784662  RESIDENCE - YARD (FRONT / BACK)    True     False  ...  45.0   
7784663                         SIDEWALK    True     False  ...  29.0   

         Community Area  FBI Code  X Coordinate Y Coordinate  Year   
0                  61.0       08B     1165074.0    1875917.0  2015  \
1                  25.0        06     1138875.0    1904869.0  2015   
2                  44.0        06           NaN          NaN  2018   
3                  21.0        18     1152037.0    1920384.0  2015   
4                  25.0       08A     1141706.0    1900086.0  2015   
...                 ...       ...           ...          ...   ...   
7784659             6.0        26     1172497.0    1919410.0  2022   
7784660            49.0        11     1179966.0    1828818.0  2022   
7784661            43.0        06     1193181.0    1859005.0  2022   
7784662            11.0        18     1140553.0    1933418.0  2022   
7784663            25.0        15     1136773.0    1899652.0  2022   

                     Updated On   Latitude  Longitude   
0        02/10/2018 03:50:01 PM  41.815117 -87.670000  \
1        02/10/2018 03:50:01 PM  41.895080 -87.765400   
2        04/06/2019 04:04:43 PM        NaN        NaN   
3        02/10/2018 03:50:01 PM  41.937406 -87.716650   
4        02/10/2018 03:50:01 PM  41.881903 -87.755121   
...                         ...        ...        ...   
7784659  01/03/2023 03:46:28 PM  41.934305 -87.641485   
7784660  01/03/2023 03:46:28 PM  41.685544 -87.616813   
778466

In [4]:
df.columns

Index(['ID', 'Case Number', 'Date', 'Block', 'IUCR', 'Primary Type',
       'Description', 'Location Description', 'Arrest', 'Domestic', 'Beat',
       'District', 'Ward', 'Community Area', 'FBI Code', 'X Coordinate',
       'Y Coordinate', 'Year', 'Updated On', 'Latitude', 'Longitude',
       'Location'],
      dtype='object')

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7784664 entries, 0 to 7784663
Data columns (total 22 columns):
 #   Column                Dtype  
---  ------                -----  
 0   ID                    int64  
 1   Case Number           object 
 2   Date                  object 
 3   Block                 object 
 4   IUCR                  object 
 5   Primary Type          object 
 6   Description           object 
 7   Location Description  object 
 8   Arrest                bool   
 9   Domestic              bool   
 10  Beat                  int64  
 11  District              float64
 12  Ward                  float64
 13  Community Area        float64
 14  FBI Code              object 
 15  X Coordinate          float64
 16  Y Coordinate          float64
 17  Year                  int64  
 18  Updated On            object 
 19  Latitude              float64
 20  Longitude             float64
 21  Location              object 
dtypes: bool(2), float64(7), int64(3), object(1

In [4]:
df.isnull().sum()

ID                           0
Case Number                  4
Date                         0
Block                        0
IUCR                         0
Primary Type                 0
Description                  0
Location Description     10381
Arrest                       0
Domestic                     0
Beat                         0
District                    47
Ward                    614848
Community Area          613476
FBI Code                     0
X Coordinate             86848
Y Coordinate             86848
Year                         0
Updated On                   0
Latitude                 86848
Longitude                86848
Location                 86848
dtype: int64

Finding out how many null values are in each column and seeing if they would impact the data. If they do not then they are worthy of being filled in statistically. If to many null values then column must be considered to be removed. 
Total entries is :7,784,664

Location Description: 10381/7784664 = 00.13% (Meaning percentage is low, fill in null values to have an accurate reading on info)

District : 47 / 7784664 = 6.03% (Percentage is low, fill in null values)

ward : 614848 / 7784664 = 7.89% (Perncentage is low, fill in null values)

Community Area : 613476 / 7784664 = 7.88% (Percentage is low, fill in null values) 

X Coordinate : 86848 / 7784664 = 1.11% (Percentage is low, fill in null values)

Y Coordinate : 86848 / 7784664 = 1.11% (Percentage is low, fill in null values) 

Latitude : 86848 / 7784664 = 1.11% (Percentage is low, fill in null values)

Longitude: 86848 / 7784664 = 1.11% (Percentage is low, fill in null values)

Location : 86848 / 7784664 = 1.11% (Percentage is low, fill in null values) 



In [3]:
df.fillna(df.mean(numeric_only=True).round(1), inplace=True)
df.isnull().sum()


ID                          0
Case Number                 4
Date                        0
Block                       0
IUCR                        0
Primary Type                0
Description                 0
Location Description    10381
Arrest                      0
Domestic                    0
Beat                        0
District                    0
Ward                        0
Community Area              0
FBI Code                    0
X Coordinate                0
Y Coordinate                0
Year                        0
Updated On                  0
Latitude                    0
Longitude                   0
Location                86848
dtype: int64

In [4]:
string_columns = df.select_dtypes(include=['object']).columns
df[string_columns] = df[string_columns].fillna(df[string_columns].mode().iloc[0])
df.isnull().sum()

ID                      0
Case Number             0
Date                    0
Block                   0
IUCR                    0
Primary Type            0
Description             0
Location Description    0
Arrest                  0
Domestic                0
Beat                    0
District                0
Ward                    0
Community Area          0
FBI Code                0
X Coordinate            0
Y Coordinate            0
Year                    0
Updated On              0
Latitude                0
Longitude               0
Location                0
dtype: int64

Now that all the data is cleaned, lets see what were working with. 
### VISUALIZATIONS 